lib/sycamore/sycamore/transforms/aggregation.py

```diff
@@ -0,0 +1,259 @@
+from typing import Callable, Optional, TYPE_CHECKING, Union
+
+from sycamore.data.document import split_data_metadata
+from sycamore.plan_nodes import UnaryNode, Node
+from sycamore.data import Document, MetadataDocument
+
+from sycamore.utils.lineage_utils import update_lineage
+from sycamore.utils.thread_local import ADD_METADATA_TO_OUTPUT, ThreadLocal
+
+if TYPE_CHECKING:
+    from ray.data import Dataset
+    import numpy as np
+
+
+class AggregationNode(UnaryNode):
+    def __init__(
+        self,
+        child: Optional[Node],
+        name: str,
+        accumulate_docs: Callable[[list[Document]], Document],
+        combine_partials: Callable[[Document, Document], Document],
+        finalize: Callable[[Document], Document],
+        group_key_fn: Callable[[Document], str] = lambda d: "nogrouping",
+        zero_factory: Callable[[], Document] = Document,
+    ):
+        super().__init__(child)
+        self._name = name
+        self._accumulate = accumulate_docs
+        self._combine = combine_partials
+        self._finalize = finalize
+        self._group_key_fn = group_key_fn
+        self._zero_factory = zero_factory
+
+    def _to_key_val(self, row):
+        doc = Document.from_row(row)
+        if isinstance(doc, MetadataDocument):
+            row["key"] = f"md-{doc.doc_id}"
+        else:
+            row["key"] = self._group_key_fn(doc)
+        return row
+
+    def _unpack(self, row):
+        import pickle
+
+        doc_n_meta = row[self._name]
+        doc = doc_n_meta["doc"]
+        meta = pickle.loads(doc_n_meta["meta"])
+        return [{"doc": doc}] + [m.to_row() for m in meta]
+
+    def execute(self, **kwargs) -> "Dataset":
+        from ray.data.aggregate import AggregateFnV2
+        import pyarrow as pa
```

Comment:
Can we keep these without shortnames? They're both only used once so the shortening doesn't save anything, and it's more confusing to read.